src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAnalysisController.java

```diff
@@ -444,34 +461,50 @@ public JsonResult updateCommonEntityInAnalysis(@PathVariable("analysisId") Long
     @ApiOperation("Upload file to attach to analysis.")
     @RequestMapping(value = "/api/v1/analysis-management/analyses/{analysisId}/upload", method = POST)
     public JsonResult<List<AnalysisFileDTO>> uploadFile(Principal principal,
-                                                        @Valid UploadFileDTO uploadFileDTO,
+                                                        @Valid UploadFilesDTO uploadFilesLinks,
                                                         @PathVariable("analysisId") Long id)
-            throws PermissionDeniedException, NotExistException, IOException {
+            throws PermissionDeniedException, NotExistException {
 
         IUser user = getUser(principal);
         T analysis = analysisService.getById(id);
-
         JsonResult<List<AnalysisFileDTO>> result = new JsonResult<>(NO_ERROR);
-
         List<AnalysisFileDTO> createdFiles = new ArrayList<>();
-
-        if (uploadFileDTO.getFile() != null) {
-            AnalysisFile createdFile = analysisService.saveFile(uploadFileDTO.getFile(), user, analysis, uploadFileDTO.getLabel(), uploadFileDTO.getExecutable(), null);
-            createdFiles.add(conversionService.convert(createdFile, AnalysisFileDTO.class));
-        } else {
-            if (StringUtils.hasText(uploadFileDTO.getLink())) {
-                AnalysisFile createdFile = analysisService.saveFile(uploadFileDTO.getLink(), user, analysis, uploadFileDTO.getLabel(), uploadFileDTO.getExecutable());
-                createdFiles.add(conversionService.convert(createdFile, AnalysisFileDTO.class));
-            } else {
-                result.setErrorCode(VALIDATION_ERROR.getCode());
-            }
-
+        Map<String, Object> validationErrors = new HashMap<>();
+        List<String> errorFileMessages = new ArrayList<>();
+        List<UploadFileDTO> files = uploadFilesLinks.getFiles().size() > 0 ? uploadFilesLinks.getFiles() : uploadFilesLinks.getLinks();
+        saveFiles(files, user, analysis, result, createdFiles, errorFileMessages);
+        if (result.getErrorCode().equals(ALREADY_EXIST.getCode())) {
```

Comment:
Couldn't we put this into the `saveFiles` function and then we could return `result` from there w/o changes by ref